### PR TITLE
Fix #186: use core data for search url completion.

### LINKED ISSUE #186: Use Core Data model for search autocomplete.
Currently it still uses firefoxes sqlite implementation.
We have it already implemented in v1 so shouldn't be hard to port it over.

### PR DESCRIPTION
**Pull request by iccub**
It's mostly a port from 1.6 implementation.

## Pull Request Checklist

- [ ] My patch has gone through review and I have addressed review comments
- [x] My patch has a standard commit message that looks like `Fix #123: This fixes the shattered coffee cup!`

- [ ] I have updated the *Unit Tests* to cover new or changed functionality
- [ ] I have updated the *UI Tests* to cover new or changed functionality
- [ ] I have marked the bug with `[needsuplift]`
- [ ] I have made sure that localizable strings use `NSLocalizableString()`

## Screenshots

If your patch includes user interface changes that you would like to suggest or that you would like UX to look at, please include them here.

## Notes for testing this patch

If useful, please leave notes for QA, explaining what this patch changes and how it can be best tested and verified.

### CODE REVIEW

File: Client/Frontend/Browser/Search/SearchLoader.swift [hunk 1]
```diff
@@ -19,10 +21,11 @@ typealias SearchLoader = _SearchLoader<AnyObject, AnyObject>
  * Shared data source for the SearchViewController and the URLBar domain completion.
  * Since both of these use the same SQL query, we can perform the query once and dispatch the results.
  */
-class _SearchLoader<UnusedA, UnusedB>: Loader<Cursor<Site>, SearchViewController> {
+class _SearchLoader<UnusedA, UnusedB>: Loader<[Site], SearchViewController> {
     fileprivate let profile: Profile
     fileprivate let urlBar: URLBarView
     fileprivate let frecentHistory: FrecentHistory
+    fileprivate var inProgress: Cancellable? = nil
```
Comment (kylehickinson) [inline Client/Frontend/Browser/Search/SearchLoader.swift]: Don't need to set it to `nil`, an optional's default is `nil`

File: Client/Frontend/Browser/Search/SearchLoader.swift [hunk 2]
```diff
@@ -49,52 +52,59 @@ class _SearchLoader<UnusedA, UnusedB>: Loader<Cursor<Site>, SearchViewController
             }
 
             if query.isEmpty {
-                load(Cursor(status: .success, msg: "Empty query"))
+                load([Site]())
```
Comment (kylehickinson) [inline Client/Frontend/Browser/Search/SearchLoader.swift]: Does this require `[Site]()`? Can it not be inferred based on the argument and just use `[]`

File: Client/Frontend/Browser/Search/SearchLoader.swift [hunk 3]
```diff
@@ -49,52 +52,59 @@ class _SearchLoader<UnusedA, UnusedB>: Loader<Cursor<Site>, SearchViewController
...
+                    if let completion = self.completionForDomain(domain) {
+                        self.urlBar.setAutocompleteSuggestion(completion)
+                        return
                     }
                 }
             }
         }
     }
+    
+    // TODO: This is not a proper frecency query, it just gets sites from the past week
```
Comment (kylehickinson) [inline Client/Frontend/Browser/Search/SearchLoader.swift]: Perhaps make a new ticket for this and reference it in the TODO comment

File: Client/Frontend/Browser/Search/SearchViewController.swift [hunk 4]
```diff
@@ -386,21 +385,20 @@ class SearchViewController: SiteTableViewController, KeyboardHelperDelegate, Loa
...
-                            cell.imageView?.backgroundColor = color
-                        }
-                    })
-                }
+            let site = data[indexPath.row]
+            if let cell = cell as? TwoLineTableViewCell {
+                let isBookmark = site.bookmarked ?? false
+                cell.setLines(site.title, detailText: site.url)
+                cell.setRightBadge(isBookmark ? self.bookmarkedBadge : nil)
+                cell.imageView!.layer.borderColor = SearchViewControllerUX.IconBorderColor.cgColor
```
Comment (kylehickinson) [inline Client/Frontend/Browser/Search/SearchViewController.swift]: I know this was just copy pasted, but is there a reason these 2 lines force-unwrap where as the rest optional-unwrap?

**Review by kylehickinson — CHANGES_REQUESTED** [review]
Looks good, just some minor nits here and there. Also are there any unit tests around the frecency stuff? It might be a good idea to add a few since we are making them `public` :)

**Comment by iccub** [comment]
Added two tests.

**Review by kylehickinson — APPROVED** [review]

**Pull request merged**